wrapper/src/main/java/software/amazon/jdbc/plugin/failover/FailoverConnectionPlugin.java

```diff
@@ -577,64 +607,129 @@ protected synchronized void failover(final HostSpec failedHost) throws SQLExcept
-    if (failedHostSpec != null && failedHostSpec.getRawAvailability() == HostAvailability.AVAILABLE) {
-      failedHost = failedHostSpec;
-    }
-    final ReaderFailoverResult result = readerFailoverHandler.failover(this.pluginService.getHosts(), failedHost);
+    try {
+      LOGGER.fine(() -> Messages.get("Failover.startReaderFailover"));
 
-    if (result != null) {
-      final SQLException exception = result.getException();
-      if (exception != null) {
-        throw exception;
+      HostSpec failedHost = null;
+      final Set<String> oldAliases = this.pluginService.getCurrentHostSpec().getAliases();
+      if (failedHostSpec != null
+          && failedHostSpec.getAvailability() == HostAvailability.AVAILABLE) {
+        failedHost = failedHostSpec;
       }
-    }
+      final ReaderFailoverResult result = readerFailoverHandler.failover(
+          this.pluginService.getHosts(), failedHost);
 
-    if (result == null || !result.isConnected()) {
-      // "Unable to establish SQL connection to reader instance"
-      processFailoverFailure(Messages.get("Failover.unableToConnectToReader"));
-      return;
-    }
+      if (result != null) {
+        final SQLException exception = result.getException();
+        if (exception != null) {
+          throw exception;
+        }
+      }
 
-    this.pluginService.setCurrentConnection(result.getConnection(), result.getHost());
+      if (result == null || !result.isConnected()) {
+        // "Unable to establish SQL connection to reader instance"
+        processFailoverFailure(Messages.get("Failover.unableToConnectToReader"));
+        this.failoverReaderFailedCounter.inc();
+        return;
+      }
 
-    this.pluginService.getCurrentHostSpec().removeAlias(oldAliases.toArray(new String[]{}));
-    updateTopology(true);
+      this.pluginService.setCurrentConnection(result.getConnection(), result.getHost());
 
-    LOGGER.fine(
-        () -> Messages.get(
-            "Failover.establishedConnection",
-            new Object[]{this.pluginService.getCurrentHostSpec()}));
+      this.pluginService.getCurrentHostSpec().removeAlias(oldAliases.toArray(new String[]{}));
+      updateTopology(true);
+
+      LOGGER.fine(
+          () -> Messages.get(
+              "Failover.establishedConnection",
+              new Object[]{this.pluginService.getCurrentHostSpec()}));
+
+      this.failoverReaderSuccessCounter.inc();
+
+    } catch (FailoverFailedSQLException ex) {
+      telemetryContext.setSuccess(false);
+      telemetryContext.setException(ex);
+      this.failoverReaderFailedCounter.inc();
+      throw ex;
+    } catch (FailoverSQLException ex) {
+      telemetryContext.setSuccess(true);
+      telemetryContext.setException(ex);
+      this.failoverReaderSuccessCounter.inc();
+      throw ex;
+    } catch (Exception ex) {
+      telemetryContext.setSuccess(false);
+      telemetryContext.setException(ex);
+      this.failoverReaderFailedCounter.inc();
+      throw ex;
```

Comment:
```suggestion
    } catch (FailoverSQLException ex) {
      telemetryContext.setSuccess(true);
      telemetryContext.setException(ex);
      this.failoverReaderSuccessCounter.inc();
      throw ex;
    } catch (Exception ex) {
      telemetryContext.setSuccess(false);
      telemetryContext.setException(ex);
      this.failoverReaderFailedCounter.inc();
      throw ex;
```

Reply:
^ @karenc-bq I'm not sure that this suggestion is correct. The point of original code is to set success flag depending on failover exception.